### ISSUE TITLE
Poetry seems to switch python version requirements for some versions of python

### ISSUE THREAD

**Issue by laurentS** [python-poetry/poetry#4499]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: debian linux unstable
- **Poetry version**: `1.1.8`, running on python 3.9.7
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: the whole content is below

## Issue

Poetry seems to modify my python version constraints for python < 3.10, when installing a single dependency in develop mode:
With the following `pyproject.toml`
```toml
[tool.poetry]
name = "my_project"
version = "0.1.0"

[tool.poetry.dependencies]
python = "^3.10"   <-- this is what matters
tap-github = {path = "../../../../../tap-github-sdk", develop = true}

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```
I get the following output for `poetry lock -vvv` (this part seems correct), note the python version constraint on both lines. The solver correctly identifies the incompatible versions.

```
The current project's Python requirement (>=3.10,<4.0) is not compatible with some of the required packages Python requirement:
    - singer-sdk requires Python >=3.6.1,<3.10, so it will not be satisfied for Python >=3.10,<4.0
```

<details>

```
Updating dependencies
Resolving dependencies...
   1: fact: ovio-data-ingestion is 0.1.0
   1: derived: ovio-data-ingestion
   1: fact: ovio-data-ingestion depends on tap-github (0.0.1)
   1: selecting ovio-data-ingestion (0.1.0)
   1: derived: tap-github (0.0.1 ../../../../../tap-github-sdk)
   1: fact: tap-github (0.0.1) depends on requests (^2.25.1)
   1: fact: tap-github (0.0.1) depends on singer-sdk (^0.3.7)
   1: selecting tap-github (0.0.1 /home/laurent/Sites/ovio/tap-github-sdk)
   1: derived: singer-sdk (>=0.3.7,<0.4.0)
   1: derived: requests (>=2.25.1,<3.0.0)
PyPI: 1 packages found for singer-sdk >=0.3.7,<0.4.0
PyPI: No release information found for requests-0.0.1, skipping
PyPI: No release information found for requests-0.12.01, skipping
PyPI: No release information found for requests-2.15.0, skipping
PyPI: 2 packages found for requests >=2.25.1,<3.0.0
   1: fact: singer-sdk (0.3.7) requires Python >=3.6.1,<3.10
   1: derived: not singer-sdk (==0.3.7)
   1: fact: no versions of singer-sdk match >0.3.7,<0.4.0
   1: conflict: no versions of singer-sdk match >0.3.7,<0.4.0
   1: !  singer-sdk (>0.3.7,<0.4.0) is partially satisfied by not  singer-sdk (0.3.7)
   1: ! which is caused by "singer-sdk (0.3.7) requires Python >=3.6.1,<3.10"
   1: ! thus: singer-sdk is forbidden
   1: fact: singer-sdk is forbidden
   1: derived: not singer-sdk (>=0.3.7,<0.4.0)
   1: conflict: tap-github (0.0.1) depends on singer-sdk (^0.3.7)
   1: ! not  singer-sdk (^0.3.7) is satisfied by not  singer-sdk (>=0.3.7,<0.4.0)
   1: ! which is caused by "singer-sdk is forbidden"
   1: ! thus: tap-github is forbidden
   1: !  tap-github (0.0.1) is satisfied by  tap-github (0.0.1)
   1: ! which is caused by "ovio-data-ingestion depends on tap-github (0.0.1)"
   1: ! thus: version solving failed
   1: Version solving took 0.202 seconds.
   1: Tried 1 solutions.

  Stack trace:

  8  ~/.poetry/lib/poetry/_vendor/py3.9/clikit/console_application.py:131 in run
      129│             parsed_args = resolved_command.args
      130│
    → 131│             status_code = command.handle(parsed_args, io)
      132│         except KeyboardInterrupt:
      133│             status_code = 1

  7  ~/.poetry/lib/poetry/_vendor/py3.9/clikit/api/command/command.py:120 in handle
      118│     def handle(self, args, io):  # type: (Args, IO) -> int
      119│         try:
    → 120│             status_code = self._do_handle(args, io)
      121│         except KeyboardInterrupt:
      122│             if io.is_debug():

  6  ~/.poetry/lib/poetry/_vendor/py3.9/clikit/api/command/command.py:171 in _do_handle
      169│         handler_method = self._config.handler_method
      170│
    → 171│         return getattr(handler, handler_method)(args, io, self)
      172│
      173│     def __repr__(self):  # type: () -> str

  5  ~/.poetry/lib/poetry/_vendor/py3.9/cleo/commands/command.py:92 in wrap_handle
       90│         self._command = command
       91│
    →  92│         return self.handle()
       93│
       94│     def handle(self):  # type: () -> Optional[int]

  4  ~/.poetry/lib/poetry/console/commands/lock.py:34 in handle
      32│         self._installer.lock(update=not self.option("no-update"))
      33│
    → 34│         return self._installer.run()
      35│

  3  ~/.poetry/lib/poetry/installation/installer.py:103 in run
      101│         local_repo = Repository()
      102│
    → 103│         return self._do_install(local_repo)
      104│
      105│     def dry_run(self, dry_run=True):  # type: (bool) -> Installer

  2  ~/.poetry/lib/poetry/installation/installer.py:235 in _do_install
      233│             )
      234│
    → 235│             ops = solver.solve(use_latest=self._whitelist)
      236│         else:
      237│             self._io.write_line("Installing dependencies from lock file")

  1  ~/.poetry/lib/poetry/puzzle/solver.py:65 in solve
       63│         with self._provider.progress():
       64│             start = time.time()
    →  65│             packages, depths = self._solve(use_latest=use_latest)
       66│             end = time.time()
       67│

  SolverProblemError

  The current project's Python requirement (>=3.10,<4.0) is not compatible with some of the required packages Python requirement:
    - singer-sdk requires Python >=3.6.1,<3.10, so it will not be satisfied for Python >=3.10,<4.0

  Because no versions of singer-sdk match >0.3.7,<0.4.0
   and singer-sdk (0.3.7) requires Python >=3.6.1,<3.10, singer-sdk is forbidden.
  So, because ovio-data-ingestion depends on tap-github (0.0.1) which depends on singer-sdk (^0.3.7), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties

    For singer-sdk, a possible solution would be to set the `python` property to "<empty>"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
```

</details>

But if I change my `pyproject.toml` to this:

```toml
[tool.poetry]
name = "my_project"
version = "0.1.0"
description = ""
authors = []

[tool.poetry.dependencies]
python = "^3.9"  <-- this is what I actually want
tap-github = {path = "../../../../../tap-github-sdk", develop = true}

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

then the output looks like this. Somehow magically, the version constraint has moved from `>=3.9` to `>=3.10`, which then becomes incompatible with the dependency. This seems to happen if I specify `^3.8` and `^3.7` as well (both "become" 3.10). If I specify `^3.6` then it becomes `>=3.6,<3.6.1 || >=3.10,<4.0`. If I instead specify by version constraint as `>=3.9`, it becomes `>=3.10`. There seems to be something wrong with version codes with a single digit for some reason (specifying any python version with 2 "decimals" seems to produce the expected output 3.10 3.11 3.12...)

Poetry then suggests a few "solutions" by making the version constraints on `singer-sdk` even tighter, or what they already are, which tells me that there's a bug somewhere in poetry.

```
The current project's Python requirement (>=3.9,<4.0) is not compatible with some of the required packages Python requirement:
    - singer-sdk requires Python >=3.6.1,<3.10, so it will not be satisfied for Python >=3.10,<4.0
```

<details>

```
Updating dependencies
Resolving dependencies...
   1: fact: ovio-data-ingestion is 0.1.0
   1: derived: ovio-data-ingestion
   1: fact: ovio-data-ingestion depends on tap-github (0.0.1)
   1: selecting ovio-data-ingestion (0.1.0)
   1: derived: tap-github (0.0.1 ../../../../../tap-github-sdk)
   1: fact: tap-github (0.0.1) depends on requests (^2.25.1)
   1: fact: tap-github (0.0.1) depends on singer-sdk (^0.3.7)
   1: selecting tap-github (0.0.1 /home/laurent/Sites/ovio/tap-github-sdk)
   1: derived: singer-sdk (>=0.3.7,<0.4.0)
   1: derived: requests (>=2.25.1,<3.0.0)
PyPI: 1 packages found for singer-sdk >=0.3.7,<0.4.0
PyPI: No release information found for requests-0.0.1, skipping
PyPI: No release information found for requests-0.12.01, skipping
PyPI: No release information found for requests-2.15.0, skipping
PyPI: 2 packages found for requests >=2.25.1,<3.0.0
   1: fact: singer-sdk (0.3.7) requires Python >=3.6.1,<3.10
   1: derived: not singer-sdk (==0.3.7)
   1: fact: no versions of singer-sdk match >0.3.7,<0.4.0
   1: conflict: no versions of singer-sdk match >0.3.7,<0.4.0
   1: !  singer-sdk (>0.3.7,<0.4.0) is partially satisfied by not  singer-sdk (0.3.7)
   1: ! which is caused by "singer-sdk (0.3.7) requires Python >=3.6.1,<3.10"
   1: ! thus: singer-sdk is forbidden
   1: fact: singer-sdk is forbidden
   1: derived: not singer-sdk (>=0.3.7,<0.4.0)
   1: conflict: tap-github (0.0.1) depends on singer-sdk (^0.3.7)
   1: ! not  singer-sdk (^0.3.7) is satisfied by not  singer-sdk (>=0.3.7,<0.4.0)
   1: ! which is caused by "singer-sdk is forbidden"
   1: ! thus: tap-github is forbidden
   1: !  tap-github (0.0.1) is satisfied by  tap-github (0.0.1)
   1: ! which is caused by "ovio-data-ingestion depends on tap-github (0.0.1)"
   1: ! thus: version solving failed
   1: Version solving took 0.104 seconds.
   1: Tried 1 solutions.

  Stack trace:

  8  ~/.poetry/lib/poetry/_vendor/py3.9/clikit/console_application.py:131 in run
      129│             parsed_args = resolved_command.args
      130│
    → 131│             status_code = command.handle(parsed_args, io)
      132│         except KeyboardInterrupt:
      133│             status_code = 1

  7  ~/.poetry/lib/poetry/_vendor/py3.9/clikit/api/command/command.py:120 in handle
      118│     def handle(self, args, io):  # type: (Args, IO) -> int
      119│         try:
    → 120│             status_code = self._do_handle(args, io)
      121│         except KeyboardInterrupt:
      122│             if io.is_debug():

  6  ~/.poetry/lib/poetry/_vendor/py3.9/clikit/api/command/command.py:171 in _do_handle
      169│         handler_method = self._config.handler_method
      170│
    → 171│         return getattr(handler, handler_method)(args, io, self)
      172│
      173│     def __repr__(self):  # type: () -> str

  5  ~/.poetry/lib/poetry/_vendor/py3.9/cleo/commands/command.py:92 in wrap_handle
       90│         self._command = command
       91│
    →  92│         return self.handle()
       93│
       94│     def handle(self):  # type: () -> Optional[int]

  4  ~/.poetry/lib/poetry/console/commands/lock.py:34 in handle
      32│         self._installer.lock(update=not self.option("no-update"))
      33│
    → 34│         return self._installer.run()
      35│

  3  ~/.poetry/lib/poetry/installation/installer.py:103 in run
      101│         local_repo = Repository()
      102│
    → 103│         return self._do_install(local_repo)
      104│
      105│     def dry_run(self, dry_run=True):  # type: (bool) -> Installer

  2  ~/.poetry/lib/poetry/installation/installer.py:235 in _do_install
      233│             )
      234│
    → 235│             ops = solver.solve(use_latest=self._whitelist)
      236│         else:
      237│             self._io.write_line("Installing dependencies from lock file")

  1  ~/.poetry/lib/poetry/puzzle/solver.py:65 in solve
       63│         with self._provider.progress():
       64│             start = time.time()
    →  65│             packages, depths = self._solve(use_latest=use_latest)
       66│             end = time.time()
       67│

  SolverProblemError

  The current project's Python requirement (>=3.9,<4.0) is not compatible with some of the required packages Python requirement:
    - singer-sdk requires Python >=3.6.1,<3.10, so it will not be satisfied for Python >=3.10,<4.0

  Because no versions of singer-sdk match >0.3.7,<0.4.0
   and singer-sdk (0.3.7) requires Python >=3.6.1,<3.10, singer-sdk is forbidden.
  So, because ovio-data-ingestion depends on tap-github (0.0.1) which depends on singer-sdk (^0.3.7), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties

    For singer-sdk, a possible solution would be to set the `python` property to ">=3.9,<3.10"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
```

</details>

**Comment by finswimmer** [comment]
Hello @laurentS,

I cannot completely follow your issue. I'm missing some example of the output you are describing.

What poetry is telling you here, is that  `singer-sdk` is not available for every python requirement of your project. If you say your project is compatible to `^3.X`, this means for any python version >=3.X but smaller 4.0. `singer-sdk` is only available for `>=3.6.1,<3.10`. So not for 3.11, 3.12 or 3.6.0.

Does this answer your question?

fin swimmer

**Comment by finswimmer** [comment]
Poetry doesn't care about your current environment when resolving dependencies. It cares about the project definition. And if the project claims to be compatible for a certain range of python versions, poetry will throw an error if you try to add a dependency which claims to be compatible only for a subset of that range.

This is fundamental different to pip. Because pip is a general purpose installation tool it cares only about the current environment.
